lib/hcrypto/evp-pkcs11.c

```diff
@@ -0,0 +1,744 @@
+                                                                        \
+    static EVP_CIPHER                                                   \
+    pkcs11_##name = {                                                   \
+        0,                                                              \
+        block_size,                                                     \
+        key_len,                                                        \
+        iv_len,                                                         \
+        flags,                                                          \
+        p11_key_init,                                                   \
+        p11_do_cipher,                                                  \
+        p11_cleanup,                                                    \
+        sizeof(struct pkcs11_cipher_ctx),                               \
+        NULL,                                                           \
+        NULL,                                                           \
+        NULL,                                                           \
+        (void *)mechanismType                                           \
+    };                                                                  \
+                                                                        \
+    const EVP_CIPHER *                                                  \
+    hc_EVP_pkcs11_##name(void)                                          \
+    {                                                                   \
+        if (p11_mech_available_p(mechanismType))                        \
+            return &pkcs11_##name;                                      \
+        else                                                            \
+            return NULL;                                                \
+    }                                                                   \
+                                                                        \
+    const EVP_CIPHER *                                                  \
+    hc_EVP_pkcs11_hcrypto_##name(void)                                  \
+    {                                                                   \
+        static const EVP_CIPHER *__cipher;                              \
+                                                                        \
+        if (__cipher == NULL) {                                         \
+            const EVP_CIPHER *cipher;                                   \
+                                                                        \
+            cipher = hc_EVP_pkcs11_ ##name();                           \
+            if (cipher == NULL)                                         \
+                cipher = hc_EVP_hcrypto_ ##name();                      \
+                                                                        \
+            (void)heim_base_exchange_pointer(&__cipher, cipher);        \
```

Comment:
This is a lot easier than using `heim_base_once` though...  It's racy but idempotent.